File: my_py_pkg/my_py_pkg/my_first_node.py
#first node in python in ros 2
#!/usr/bin/env python3
import rclpy 
from rclpy.node import Node  

class MyNode(Node):
    
    def __init__(self): #contructor
        super().__init__("Py_test_first_N_OOP") # name of the node 

        self.counter_ = 0 #add a counter
        
        self.get_logger().info("Hello this Is first python node")#print something

        self.create_timer(1 , self.timer_callback)#create a timer ie how many time spam it will run
    
    
    # now write a function and use this timer for 0.5 hz && use the counter in the func
        
    def timer_callback(self): ## take care for intendation 
       
        self.counter_ += 1

        self.get_logger().info ( " Helloo " + str(self.counter_) )



def main (args =None): #init tos communication

    rclpy.init(args =args ) # Always write this too start ros communication

    node = MyNode()#call it from the class above 
    
    
    
    rclpy.spin(node)# program will not end



    rclpy.shutdown()#exit the node 


if __name__ == "__main__":
    main()

File: my_cpp_pkg/src/my_first_node.cpp
#include "rclcpp/rclcpp.hpp"
//use shift +alt+p c/c++:config and add ros/humble/include 

using namespace std ;



// everthing you write will be in the node class and change name accordingly too usage

class MyNode: public  rclcpp::Node // take care of capital letters

{ // name of the node 

public:
    MyNode(): Node ("cpp_test"),counter_(0) // Constructor node wíth the node name 
    {

       

        RCLCPP_INFO(this-> get_logger(), "Hello Cpp node"); // print  somethimh


        timer_ = this-> create_wall_timer(std::chrono::seconds(1), // create a timer

                                            std ::bind(&MyNode::timerCallback,this )); // call the function callnack

    }




private : 
    // add a timer (callback) so it can give the print with desired feq
    void timerCallback() {

        counter_++;// increment


        RCLCPP_INFO(this-> get_logger(), "Hello %d " , counter_) ; // print in the call back function
    }

        rclcpp::TimerBase::SharedPtr timer_; // declare timer

        int counter_;// declare counter 



};
    




int main(int argc , char **argv){

    rclcpp::init(argc , argv);// init ros2 communication




    auto node = make_shared<MyNode>(); // create a node (its a shared pointer to node)!!!!!

   

    
    rclcpp::spin(node); // stop the program with ctrl +c before that function will be workinh
    rclcpp::shutdown(); // it will shutdown when we press the c and kil the node 





    return 0;
}